### PR TITLE
add support for alpha channel

### PR DESCRIPTION
**Pull request by Mottie**
Support rgba to hex 8-digit format, see https://github.com/sindresorhus/rgb-hex-cli/issues/1.

### CODE REVIEW

File: test.js [hunk 1]
```diff
@@ -6,4 +6,10 @@ test(t => {
 	t.is(m(65, 131, 196), '4183c4');
 	t.is(m('255 154 253'), 'ff9afd');
 	t.is(m('rgb(40, 42, 54)'), '282a36');
+	/* eslint no-floating-decimal:0 */
+	t.is(m(0, 0, 0, .5), '00000080');
```
Comment (sindresorhus) [inline test.js]: `.5` => `0.5`. There's no point in testing without a `0` as the module doesn't see anything different here.

File: index.js [hunk 2]
```diff
@@ -1,12 +1,13 @@
 'use strict';
 /* eslint-disable no-mixed-operators */
-module.exports = (red, green, blue) => {
+module.exports = (red, green, blue, alpha) => {
 	if (typeof red === 'string') {
-		const res = red.match(/\b\d{1,3}\b/g).map(Number);
+		const res = red.match(/(0?\.?\d{1,3}%?)\b/g).map(Number);
```
Comment (sindresorhus) [inline index.js]: Would be clearer when reading the regex if `%?` were outside the capture group.

**Comment by Mottie** [comment]
Done :grin:

File: index.js [hunk 3]
```diff
@@ -18,5 +19,17 @@ module.exports = (red, green, blue) => {
 		throw new TypeError('Expected three numbers below 256');
 	}
 
-	return ((blue | green << 8 | red << 16) | 1 << 24).toString(16).slice(1);
+	if (typeof alpha === 'number') {
+		if (alpha < 1 && alpha >= 0) {
+			alpha = Math.round(256 * alpha).toString(16);
```
Comment (Qix-) [inline index.js]: There's no need for a toString operation. Just shift the color bytes by another 8 and tack on the alpha byte at the end.

File: index.js [hunk 4]
```diff
@@ -18,5 +19,17 @@ module.exports = (red, green, blue) => {
...
+		} else if (alpha > 1 && alpha < 100) {
+			alpha = Math.round(256 * alpha / 100).toString(16);
```
Comment (Qix-) [inline index.js]: I always hate this logic. Stick to a single representation of alpha (I'm always in the camp of 0.0 - 1.0). @sindresorhus might have more to say, but this causes problems, say, if I'm `for` looping from 0 to 50 with a step of 0.1. My outputs are going to be crazy for the first few iterations.

In my opinion, this is trying to get _too_ fancy with your implementation.

File: index.js [hunk 5]
```diff
@@ -18,5 +19,17 @@ module.exports = (red, green, blue) => {
...
+		} else {
+			throw new TypeError('Expected alpha value as a decimal or percentage');
+		}
+	} else {
+		alpha = '';
```
Comment (Qix-) [inline index.js:35]: Why a string? This doesn't have to be a string. :)

File: index.js [hunk 6]
```diff
@@ -18,5 +19,17 @@ module.exports = (red, green, blue) => {
...
+	}
+
+	return ((blue | green << 8 | red << 16) | 1 << 24).toString(16).slice(1) + alpha;
```
Comment (Qix-) [inline index.js:38]: I would calculate this first, then determine if alpha is there and if so shift this 8 bytes and add alpha.

**Comment by Qix-** [comment]
Should probably specify in the readme that this is an RGBA implementation and not ARGB. That distinction matters.

File: index.js (hunk 4 above)
Comment (Mottie) [reply index.js]: You're right! I'll fix that.

File: index.js (hunk 3 above)
Comment (Mottie) [reply index.js]: I tried to not use strings, but we run into issues when working with more than 32 bits

``` js
let red = 0,
  green = 0,
  blue = 0,
  result = ((blue | green << 8 | red << 16) | 1 << 24).toString(16);
// result = '1000000'; <- the "1" is removed by the slice(1)

```

Now if I didn't use strings for the alpha we're beyond 32 bits and the value becomes zero:

``` js
let red = 0,
  green = 0,
  blue = 0,
  rgb = (blue | green << 8 | red << 16) | 1 << 24,
  alpha = 100;
result = (rgb << 8 | Math.round(256 * alpha / 100)).toString(16);
// result = '100';
```

So, I think using strings keeps everything readable, and it works.

File: index.js (hunk 5 above)
Comment (Mottie) [reply index.js:35]: I like my string... my precious!

**Comment by Mottie** [comment]
I shouldn't have been using `256` in the calculations .

**Comment by Qix-** [comment]
Still not convinced on all of the string operations going on as they are needless...

**Comment by Qix-** [comment]
Also the following will break your implementation.

``` js
for (var i = 0; i < 200; i += 0.1) {
    console.log(rgbHex(255, 0, 0, i));
}
```

**Comment by Mottie** [comment]
[Demo](https://jsfiddle.net/Mottie/85qpd1bw/) of the latest code.

---

I guess this should be discussed in another issue, but rgb values can also be percentages - https://drafts.csswg.org/css-color/#rgb-functions - although all three values need to be of the same type, number (0-255) or percentage (0-100%)

**Comment by sindresorhus** [comment]
> Demo of the latest code.

@Qix- ⬆️ You approve?

## 

> I guess this should be discussed in another issue, but rgb values can also be percentages - https://drafts.csswg.org/css-color/#rgb-functions - although all three values need to be of the same type, number (0-255) or percentage (0-100%)

Wow. I've been doing CSS for so many years and I didn't know. I think we can safely ignore that edge case.

**Comment by Mottie** [comment]
I also made [this demo](https://jsfiddle.net/Mottie/89wwwhef/)... it works with percentage color settings, but doesn't check if all three are of the same type. It also doesn't use any fancy bitwise operators, but @Qix- might not like all the string operations :wink:

File: readme.md [hunk 7]
```diff
@@ -1,6 +1,6 @@
 # rgb-hex [![Build Status](https://travis-ci.org/sindresorhus/rgb-hex.svg?branch=master)](https://travis-ci.org/sindresorhus/rgb-hex)
 
-> Convert RGB color to HEX
+> Convert RGB(A) color to HEX
```
Comment (sindresorhus) [inline readme.md:3]: You also need to update package.json

**Comment by sindresorhus** [comment]
@Mottie Thanks! :) Would you mind doing a PR on `rgb-hex-cli` adding support for it there too? Mainly adding it to the help output.

**Comment by Mottie** [comment]
Done. See [PR #3](https://github.com/sindresorhus/rgb-hex-cli/pull/3).

**Pull request merged**